ibis/common/tests/test_dispatch.py

```diff
@@ -0,0 +1,122 @@
+import collections
+import decimal
+
+from ibis.common.dispatch import lazy_singledispatch
+
+
+def test_lazy_singledispatch():
+    @lazy_singledispatch
+    def foo(x):
+        """A docstring."""
+        return "base result"
+
+    @foo.register(int)
+    def _(x):
+        return x + 1
+
+    @foo.register(float)
+    def _(x):
+        return x - 1
+
+    @foo.register((tuple, list))
+    def _(x):
+        return tuple(foo(i) for i in x)
+
+    class Bar:
+        pass
+
+    assert foo.__name__ == "foo"
+    assert foo.__doc__ == "A docstring."
+
+    assert foo(1) == 2
+    assert foo(1.0) == 0.0
+    assert foo((1, 2.0)) == (2, 1.0)
+    assert foo([1, 2.0]) == (2, 1.0)
+    assert foo("a string") == "base result"
+
+
+def test_lazy_singledispatch_extra_args():
+    @lazy_singledispatch
+    def foo(*args, **kwargs):
+        pass
+
+    @foo.register(int)
+    def _(a, b, c=2):
+        return a + b + c
+
+    @foo.register(float)
+    def _(a, b, c=2):
+        return a - b - c
+
+    assert foo(1, 2) == 5
+    assert foo(1.0, 2) == -3.0
+    assert foo(1, 2, c=3) == 6
+
+
+def test_lazy_singledispatch_lazy_loader():
+    @lazy_singledispatch
+    def foo(a):
+        return a
+
+    @foo.lazy_loader("decimal")
```

Comment:
Just wondering, could we directly register the fully qualified symbol as a string instead of adding a lazy loader?

```py
@foo.lazy_loader("decimal.Decimal")
def inc_decimal(a):
    return a + 1
```

Reply:
We _could_ do that, but in practice we'd have to register all types from a lazily loaded module all at the same time to ensure subclasses were handled properly by the dispatch table (effectively recreating the pattern we have now, just hiding it from the user).

If you think a single `foo.register` method that takes in a string type representation is more ergonomic, then I'm happy to change it, it'll just require some more code to make work.

Reply:
Yes, please do :)

Reply:
done.

Reply:
Thanks!